Source/com/drew/metadata/mp4/boxes/ItemListBox.java

```diff
@@ -0,0 +1,145 @@
+                String typeName = reader.getString(4);//data
+                totalRead += 16;
+                if ("data".equals(typeName)) {
+                    reader.skip(8);//not sure what these are
+                    totalRead += 8;
+                    int toRead = (int) fieldLen - 16;
+                    if (toRead <= 0) {
+                        //log?
+                        return;
+                    }
+                    if ("covr".equals(fieldName)) {
+                        //covr can be an image file, e.g. png or jpeg
+                        //skip this for now
+                        reader.skip(toRead);
+                    } else if ("cpil".equals(fieldName)) {
+                        int compilationId = (int) reader.getByte();
+//                    metadata.set(XMPDM.COMPILATION, compilationId);
+                    } else if ("trkn".equals(fieldName)) {
+                        if (toRead == 8) {
+                            long numA = reader.getUInt32();
+                            long numB = reader.getUInt32();
+//                        metadata.set(XMPDM.TRACK_NUMBER, (int)numA);
+                        } else {
+                            //log
+                            reader.skip(toRead);
+                        }
+                    } else if ("disk".equals(fieldName)) {
+                        int a = reader.getInt32();
+                        short b = reader.getInt16();
+//                    metadata.set(XMPDM.DISC_NUMBER, a);
+                    } else {
+                        String val = reader.getString(toRead);
+                        if (itemMap.containsKey(fieldName)) {
+                            ItemListEntry entry = itemMap.get(fieldName);
+                            entry.setVal(val);
+                            itemMap.put(fieldName, entry);
+                        }
+//                    try {
+//                        addMetadata(fieldName, val);
+//                    } catch (SAXException e) {
+//                        //need to punch through IOException catching in MP4Reader
+//                        throw new RuntimeSAXException(e);
+//                    }
+                    }
+
+                    totalRead += toRead;
+                } else {
+                    int toSkip = (int) recordLen - 16;
+                    if (toSkip <= 0) {
+                        //log?
+                        return;
+                    }
+                    reader.skip(toSkip);
+                    totalRead += toSkip;
+                }
+            }
+        } catch (EOFException e) {
+            System.out.println("EOF reached");
+        }
+    }
+
+    public void addMetadata(Mp4MetaBoxDirectory directory)
+    {
+        for (Map.Entry<String, ItemListEntry> entry : itemMap.entrySet()) {
+            if (entry.getValue().getVal() != null) {
+                directory.setString(entry.getValue().getTag(), entry.getValue().getVal());
```

Comment:
Here, we iterate over the entire itemMap that we have been populating.

If the value isn't null, then we add it to the directory.

There are a couple of issues as is:
1. Some values will require additional processing. This will likely just entail some more conditionals.
2. Not all values will be strings. Up to author on implementation of this.